File: typescript-examples/starter-crawler/api/crawl.ts
import { BrowserContext, Page } from "playwright";
import { goToUrl } from "@intuned/browser";

interface Params {
  url: string;
}

export default async function handler(
  params: Params,
  page: Page,
  _context: BrowserContext
) {
  await goToUrl({ page, url: params.url });

  const baseHost = new URL(params.url).host;

  const links = await page.$$eval("a[href]", (anchors) =>
    anchors
      .map((a) => (a as HTMLAnchorElement).href)
      .filter((h) => h.startsWith("http"))
  );

  const unique = Array.from(new Set(links));
  const internal = unique.filter((h) => {
    try {
      return new URL(h).host === baseHost;
    } catch {
      return false;
    }
  });

  return {
    url: params.url,
    totalLinks: unique.length,
    internalLinks: internal.length,
    links: internal.slice(0, 50),
  };
}

File: python-examples/starter-crawler/api/crawl.py
"""
Minimal native crawler starter: visit a URL and extract internal links.

This is a one-step, single-depth version of the full native-crawler template.
"""

from typing import TypedDict
from urllib.parse import urljoin, urlparse

from intuned_browser import go_to_url
from playwright.async_api import BrowserContext, Page


class Params(TypedDict, total=False):
    url: str
    max_links: int


async def automation(
    page: Page,
    params: Params,
    context: BrowserContext | None = None,
    **_kwargs,
):
    url = params.get("url")
    if not url:
        return {"success": False, "error": "url parameter is required"}

    max_links = params.get("max_links", 20)
    base_domain = urlparse(url).netloc

    await go_to_url(page, url)
    title = await page.title()

    hrefs = await page.eval_on_selector_all(
        "a[href]", "els => els.map(e => e.getAttribute('href'))"
    )

    links: list[str] = []
    seen: set[str] = set()
    for href in hrefs:
        if not href:
            continue
        absolute = urljoin(url, href).split("#")[0]
        if urlparse(absolute).netloc != base_domain:
            continue
        if absolute in seen:
            continue
        seen.add(absolute)
        links.append(absolute)
        if len(links) >= max_links:
            break

    return {
        "success": True,
        "url": url,
        "title": title,
        "links": links,
        "count": len(links),
    }
